src/include/daos_types.h

```diff
@@ -832,44 +840,116 @@ enum {
+	char				*dai_who;
 };
 
-struct daos_acl_entry {
-	/** DAOS_ACL_USER/GROUP/MASK/OTHER */
-	uint16_t		ae_tag;
-	/** permissions DAOS_PC_RO/RW */
-	uint16_t		ae_perm;
-	/** uid or gid, meaningful only when ae_tag is DAOS_ACL_USER/GROUP */
-	uint32_t		ae_id;
+/**
+ * Type of identity provided. Special types OWNER, OWNER_GROUP and EVERYONE
+ * should not provide an identity table.
+ */
+enum daos_acl_principal_type {
+	DAOS_ACL_OWNER,		/** Owner of the object */
+	DAOS_ACL_USER,		/** Individual user */
+	DAOS_ACL_OWNER_GROUP,	/** Owning group */
+	DAOS_ACL_GROUP,		/** Group */
+	DAOS_ACL_EVERYONE	/** Anyone else */
 };
 
-struct daos_acl {
-	/** number of acl entries */
-	uint32_t		 da_nr;
+/**
+ * Bits representing access types to set permissions for
+ */
+enum daos_acl_access_type {
+	DAOS_ACL_ACCESS_ALLOW = (1U << 0),	/** allow access */
+	DAOS_ACL_ACCESS_AUDIT = (1U << 1),	/** log the access for review */
+	DAOS_ACL_ACCESS_ALARM = (1U << 2)	/** notify of the access */
+};
+
+/**
+ * Bits representing access flags
+ */
+enum daos_acl_flags {
+	/** This represents a group, not a user */
+	DAOS_ACL_FLAG_GROUP		= (1U << 0),
+	/** Containers should inherit access controls from this pool */
+	DAOS_ACL_FLAG_POOL_INHERIT	= (1U << 1),
+	/** Audit/alarm should occur on failed access */
+	DAOS_ACL_FLAG_ACCESS_FAIL	= (1U << 2),
+	/** Audit/alarm should occur on successful access */
+	DAOS_ACL_FLAG_ACCESS_SUCCESS	= (1U << 3)
+};
+
+/**
+ * Bits representing the specific permissions that may be set
+ */
+enum daos_acl_permissions {
+	DAOS_ACL_PERMISSIONS_READ	= (1U << 0),
+	DAOS_ACL_PERMISSIONS_WRITE	= (1U << 1)
+};
+
+/**
+ * Table representing a specific ACL entry.
+ * Length in the header indicates the length of this table, plus the companion
+ * identity table if present.
+ */
+struct daos_acl_entry_table {
+	struct daos_acl_table_header	dat_header;
+	/** daos_acl_principal_type */
+	uint16_t			dat_principal_type;
+	/** Bitmap of daos_acl_access_type */
+	uint16_t			dat_access_types;
+	/** Bitmap of daos_acl_flags */
+	uint16_t			dat_access_flags;
+	/** for 64-bit alignment */
+	uint16_t			dat_reserv;
+	/** Bitmap of daos_acl_permissions for the ALLOW access */
+	uint64_t			dat_allow_perms;
+	/** Bitmap of daos_acl_permissions for AUDIT access */
+	uint64_t			dat_audit_perms;
+	/** Bitmap of daos_acl_permissions for ALARM access */
+	uint64_t			dat_alarm_perms;
+};
+
+/**
+ * Top level header of the ACL tables, followed by the flat table of entries.
+ * This is expected to be a contiguous blob of data, which is walked by
+ * inspecting the length of each entry.
+ */
+struct daos_acl_table {
+	/** overall length of entries list in bytes */
+	uint32_t			 da_length;
 	/** reserved for future usage (for 64 bits alignment now) */
-	uint32_t		 da_reserv;
-	/** acl entries array */
-	struct daos_acl_entry	*da_entries;
+	uint32_t			 da_reserv;
+	/** flat table of variable-length ACL entries */
+	struct daos_acl_entry_table	*da_entries;
```

Comment:
As I think about it, I think there is still value in having the length in bytes in the daos_acl - we need the total length when converting the blob to an IOV for example. Otherwise we're stuck traversing the entire list to determine the length, since each entry is variable length. Number of entries may not help us much for the same reason - we can't walk the list of entries like a normal array.